src/conn.rs

```diff
@@ -451,9 +458,56 @@ impl Queryable for Store {
 }
 
 impl Syncable for Store {
+    // TODO refactor this out of there, somehow!
     fn sync(&mut self, server_uri: &String, user_uuid: &String) -> Result<()> {
         let uuid = Uuid::parse_str(&user_uuid)?;
-        Ok(Syncer::flow(&mut self.sqlite, server_uri, &uuid)?)
+        let mut tx_list = None;
+        {
+            let mut db_tx = self.sqlite.transaction()?;
+            if let Some(list) = Syncer::flow(&mut db_tx, server_uri, &uuid)? {
```

Comment:
If you implement `flow` on `InProgress`, you get the transaction for free… the only tricky bit is that you'd ideally like to not do any work (including establishing an exclusive transaction!) if there's no data to write.